File: VisualBasicConsoleApp/Intertop.vb
Imports Microsoft.FSharp.Collections
Imports System.Runtime.CompilerServices

Public Module Intertop

    <Extension()>
    Public Function ToFSharpList(Of T)(input As List(Of T)) As FSharpList(Of T)
        Return CreateFSharpList(input, 0)
    End Function

    Private Function CreateFSharpList(Of T)(input As List(Of T), index As Integer) As FSharpList(Of T)
        If index >= input.Count Then
            Return FSharpList(Of T).Empty
        End If

        Return FSharpList(Of T).Cons(input(index), CreateFSharpList(input, index + 1))
    End Function

End Module

File: CSharpConsoleApp/Intertop.cs
﻿using Microsoft.FSharp.Collections;
using System.Collections.Generic;

namespace CSharpConsoleApp
{
    public static class Interop
    {
        public static FSharpList<T> ToFSharpList<T>(this IList<T> input)
        {
            return CreateFSharpList(input, 0);
        }

        private static FSharpList<T> CreateFSharpList<T>(IList<T> input, int index)
        {
            if (index >= input.Count)
            {
                return FSharpList<T>.Empty;
            }
            else
            {
                return FSharpList<T>.Cons(input[index], CreateFSharpList(input, index + 1));
            }
        }
    }
}
